### PR TITLE
Make all_messages_delivered_up_to take tracked_chains into account.

### PR DESCRIPTION
**Pull request by afck**
## Motivation

https://github.com/linera-io/linera-protocol/pull/2035 introduced a list of "tracked chains": In the local node, messages to non-tracked chains are not delivered and stay in the outbox.

However, `all_messages_delivered_up_to` doesn't take that into account yet, so if a chain is sending a message to a non-tracked chain, it would never return `true`, and if `wait_for_outgoing_messages` is `true`, `handle_certificate` could wait forever.

## Proposal

Return `true` from `all_messages_delivered_up_to` if all messages to _tracked_ chains have been delivered, ignoring remaining outbox entries for untracked chains.

TBD: This makes `wait_for_outgoing_messages` async and potentially much slower. We could instead add `tracked_outbox_counters` that count only messages to tracked chains. This would need to be updated when the set of tracked chains changes, however.

## Test Plan

TBD: I should add a regression test for this.

## Release Plan

- These changes should be backported to the latest `devnet` branch, then
    - be released in a new SDK,
- These changes should be backported to the latest `testnet` branch, then
    - be released in a new SDK,

## Links

- Tracked chains PR: https://github.com/linera-io/linera-protocol/pull/2035
- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-chain/src/chain.rs [hunk 1]
```diff
@@ -437,21 +437,6 @@ where
         Ok(true)
     }
 
-    /// Returns true if there are no more outgoing messages in flight up to the given
-    /// block height.
-    pub fn all_messages_delivered_up_to(&mut self, height: BlockHeight) -> bool {
-        tracing::debug!(
-            "Messages left in {:.8}'s outbox: {:?}",
-            self.chain_id(),
-            self.outbox_counters.get()
```
Comment (ma2bd) [inline linera-chain/src/chain.rs:446]: Could we instead increment the counters right away for the messages to untracked chains?

Comment (ma2bd) [reply linera-chain/src/chain.rs:446]: or just what you said: `tracked_outbox_counters`
perhaps we can also have: `untracked_outbox_counters` to support future changes in the tracked chains

Comment (afck) [reply linera-chain/src/chain.rs:446]: Actually, maybe the `outbox_counters` just need to refer only to the tracked chains. When those change, the counters have to be recomputed from the outboxes themselves anyway.

Comment (afck) [reply linera-chain/src/chain.rs:446]: …but then I wonder if we should move it from the (persisted) chain state view to the worker state?

Comment (ma2bd) [reply linera-chain/src/chain.rs:446]: This works for me!

Comment (ma2bd) [reply linera-chain/src/chain.rs:446]: (now that chain states are long lived)

Comment (afck) [reply linera-chain/src/chain.rs:446]: It feels error-prone to remove `outbox_counters` so far from the outboxes in the code. I'm now wondering whether we should rather add `tracked_chains` to the chain state? That way we can maintain the counters only for the tracked chains, and we can update them if the tracked chains change?

Another option is to remove the counters entirely: In practice, I'd expect the outboxes for the tracked chains to usually be empty?

Comment (afck) [reply linera-chain/src/chain.rs:446]: But to quickly fix the issue, I think this is by far the easiest approach.

File: linera-chain/src/chain.rs [hunk 2]
```diff
@@ -437,21 +437,6 @@ where
         Ok(true)
     }
 
-    /// Returns true if there are no more outgoing messages in flight up to the given
-    /// block height.
-    pub fn all_messages_delivered_up_to(&mut self, height: BlockHeight) -> bool {
```
Comment (ma2bd) [inline linera-chain/src/chain.rs:442]: nit: Can we keep this function as it is here? (until we decide what to do with `outbox_counters`)

File: linera-core/src/chain_worker/state/mod.rs [hunk 3]
```diff
@@ -426,6 +426,44 @@ where
...
+            "Messages left in {:.8}'s outbox: {:?}",
+            self.chain_id(),
+            self.chain.outbox_counters.get()
+        );
+        let Some((key, _)) = self.chain.outbox_counters.get().first_key_value() else {
+            return Ok(true);
+        };
+        if key > &height {
+            return Ok(true);
+        }
```
Comment (ma2bd) [inline linera-core/src/chain_worker/state/mod.rs:445]: nit: Can we call `self.chain.all_messages_delivered_up_to` first, then continue with your new code? Perhaps this function can be called `all_messages_delivered_or_intentionally_dropped_up_to` or something like that.

**Review by ma2bd — APPROVED** [review]
Thanks for the fix!

File: linera-chain/src/chain.rs (hunk 2 above)
Comment (ma2bd) [reply linera-chain/src/chain.rs:442]: The `outbox_counters` work great for validators but create a memory leak for clients with tracked chains.

File: linera-core/src/chain_worker/state/mod.rs [hunk 4]
```diff
@@ -426,6 +426,44 @@ where
...
+        if let Some(tracked_chains) = self.tracked_chains.as_ref() {
+            let mut targets = self.chain.outboxes.indices().await?;
+            {
+                let tracked_chains = tracked_chains
+                    .read()
+                    .expect("Panics should not happen while holding a lock to `tracked_chains`");
```
Comment (ma2bd) [inline linera-core/src/chain_worker/state/mod.rs:451]: "while waiting for a lock"? Perhaps `unwrap` is just fine for locks?

**Pull request merged**